pkg/testttp/http.go

```diff
@@ -10,76 +12,94 @@ import (
 type Req struct {
-	addr    string
-	method  string
-	body    io.Reader
-	headers []string
+	t testing.TB
+
+	req *http.Request
 }
 
 // HTTP runs creates a request for an http call.
-func HTTP(method, addr string, body io.Reader) *Req {
+func HTTP(t testing.TB, method, addr string, body io.Reader) *Req {
 	return &Req{
-		addr:   addr,
-		method: method,
-		body:   body,
+		t:   t,
+		req: httptest.NewRequest(method, addr, body),
 	}
 }
 
 // Delete creates a DELETE request.
-func Delete(addr string) *Req {
-	return HTTP(http.MethodDelete, addr, nil)
+func Delete(t testing.TB, addr string) *Req {
+	return HTTP(t, http.MethodDelete, addr, nil)
 }
 
 // Get creates a GET request.
-func Get(addr string) *Req {
-	return HTTP(http.MethodGet, addr, nil)
+func Get(t testing.TB, addr string) *Req {
+	return HTTP(t, http.MethodGet, addr, nil)
 }
 
 // Patch creates a PATCH request.
-func Patch(addr string, body io.Reader) *Req {
-	return HTTP(http.MethodPatch, addr, body)
+func Patch(t testing.TB, addr string, body io.Reader) *Req {
+	return HTTP(t, http.MethodPatch, addr, body)
+}
+
+// PatchJSON creates a PATCH request with a json encoded body.
+func PatchJSON(t testing.TB, addr string, v interface{}) *Req {
+	return HTTP(t, http.MethodPatch, addr, mustEncodeJSON(t, v))
 }
 
 // Post creates a POST request.
-func Post(addr string, body io.Reader) *Req {
-	return HTTP(http.MethodPost, addr, body)
+func Post(t testing.TB, addr string, body io.Reader) *Req {
+	return HTTP(t, http.MethodPost, addr, body)
+}
+
+// PostJSON returns a POST request with a json encoded body.
+func PostJSON(t testing.TB, addr string, v interface{}) *Req {
+	return Post(t, addr, mustEncodeJSON(t, v))
 }
 
 // Put creates a PUT request.
-func Put(addr string, body io.Reader) *Req {
-	return HTTP(http.MethodPut, addr, body)
+func Put(t testing.TB, addr string, body io.Reader) *Req {
+	return HTTP(t, http.MethodPut, addr, body)
 }
 
-// Headers allows the user to set headers on the http request.
-func (r *Req) Headers(k, v string, rest ...string) *Req {
-	r.headers = append(r.headers, k, v)
-	r.headers = append(r.headers, rest...)
-	return r
+// PutJSON creates a PUT request with a json encoded body.
+func PutJSON(t testing.TB, addr string, v interface{}) *Req {
+	return HTTP(t, http.MethodPut, addr, mustEncodeJSON(t, v))
 }
 
 // Do runs the request against the provided handler.
 func (r *Req) Do(handler http.Handler) *Resp {
-	req := httptest.NewRequest(r.method, r.addr, r.body)
 	rec := httptest.NewRecorder()
 
-	for i := 0; i < len(r.headers); i += 2 {
-		if i+1 >= len(r.headers) {
-			break
-		}
-		k, v := r.headers[i], r.headers[i+1]
-		req.Header.Add(k, v)
-	}
-
-	handler.ServeHTTP(rec, req)
+	handler.ServeHTTP(rec, r.req)
 
 	return &Resp{
-		Req: req,
+		t:   r.t,
+		Req: r.req,
 		Rec: rec,
 	}
 }
 
+// Headers allows the user to set headers on the http request.
+func (r *Req) Headers(k, v string, rest ...string) *Req {
+	headers := append(rest, k, v)
```

Comment:
Why don't you use a map, looks like a little hacky.

Reply:
maps makes things unique, don't want to do that here. You can have multiple header values. This design is intentional.

Reply:
The key should unique, right?

Reply:
@kelwang sorry for the earlier reply, let that in haste, was rushing out the door. Here is better answer 😬.

When you are using this pkg in a test, the vast majority of hte time (in my experience), has been to add a single header. This design allows for that most common use to be super dead simple:

```go
testttp.
     Post(...).
     Header("Content-Type", "application/json").
      ...  more stuffs
```

if I wnat to add additional headers I still can, but I dont' ahve to sacrifice the ease of using this API to do that. I can simply do the following:

```go
testttp.
     Post(...).
     Header(
         "Content-Type", "application/json",
         "Accept", "text/yaml",
      ).
      ...  more stuffs
```

The signature is a pattern I picked up from Dave Cheney(I'll try and see if I can pull up that blog post he shared about this). It's a means of providing a type safe API with compile time errors when at least 1 arg is required and the rest are optional. Here's an example to illustrate:

```go
func (r *Req) Headers(headerPairs ...string) *Req {...}

// no compile time help here
testtp.Post(...).Headers()... other stuffs 

// k,v estaglish input pattern (meaning they are pairs of k,v)
func (r *Req) Headers(k, v string, rest ...string) *Req {...}

// compile time error here
testtp.Post(...).Headers()... other stuffs 

```